File: skills_manager/scanner.py
"""Skill filesystem scanner."""
from __future__ import annotations

from pathlib import Path
from typing import Any

from . import adapters, store


def classify_entry(path: Path) -> dict[str, Any]:
    item: dict[str, Any] = {"path": str(path), "name": path.name}
    try:
        if path.is_symlink():
            target = Path(path.readlink())
            item["link_target"] = str(target)
            resolved = path.resolve(strict=False)
            if not path.exists():
                item["type"] = "broken_symlink"
                return item
            item["resolved"] = str(resolved)
            item["type"] = "symlink_skill" if store.is_skill_dir(path) else "symlink_non_skill"
        elif path.is_dir():
            item["type"] = "skill_dir" if (path / "SKILL.md").is_file() else "missing_skill_md"
        else:
            item["type"] = "file"
        if item["type"] in {"skill_dir", "symlink_skill"}:
            item["content_hash"] = store.content_hash(path)
    except OSError as exc:
        item["type"] = "error"
        item["error"] = str(exc)
    return item


def scan_dir(path: Path) -> dict[str, Any]:
    result: dict[str, Any] = {"path": str(path), "exists": path.exists(), "entries": []}
    if not path.exists():
        return result
    if not path.is_dir():
        result["error"] = "not a directory"
        return result
    result["entries"] = [classify_entry(p) for p in sorted(path.iterdir(), key=lambda x: x.name)]
    return result


def scan(project: str | Path | None = None) -> dict[str, Any]:
    locations: dict[str, Path] = {
        "inbox": store.inbox_dir(),
        "store": store.skills_root(),
        "claude_global": adapters.adapter("claude").global_dir(),
        "codex_global": adapters.adapter("codex").global_dir(),
    }
    if project:
        locations["claude_project"] = adapters.adapter("claude").project_dir(project)
        locations["codex_project"] = adapters.adapter("codex").project_dir(project)
    out: dict[str, Any] = {"locations": {name: scan_dir(path) for name, path in locations.items()}}
    names: dict[str, list[str]] = {}
    hashes: dict[str, list[str]] = {}
    for loc_name, loc in out["locations"].items():
        for entry in loc.get("entries", []):
            if entry.get("type") in {"skill_dir", "symlink_skill"}:
                names.setdefault(entry["name"], []).append(f"{loc_name}:{entry['path']}")
                if entry.get("content_hash"):
                    hashes.setdefault(entry["content_hash"], []).append(f"{loc_name}:{entry['path']}")
    out["duplicates"] = {
        "names": {k: v for k, v in names.items() if len(v) > 1},
        "content_hashes": {k: v for k, v in hashes.items() if len(v) > 1},
    }
    return out

File: src/core/scanner.ts
import { lstat, realpath, readdir, readlink, stat } from "node:fs/promises";
import { basename, join, resolve } from "node:path";
import { adapter } from "./adapters.js";
import { contentHash, isSkillDir } from "./store.js";
import { inboxDir, skillsRoot, type Env } from "./paths.js";

export type ScanEntry = {
  path: string;
  name: string;
  type: "broken_symlink" | "symlink_skill" | "symlink_non_skill" | "skill_dir" | "missing_skill_md" | "file" | "error";
  link_target?: string;
  resolved?: string;
  content_hash?: string;
  error?: string;
};

export type ScanDirResult = {
  path: string;
  exists: boolean;
  entries: ScanEntry[];
  error?: string;
};

export async function classifyEntry(path: string): Promise<ScanEntry> {
  const item: ScanEntry = { path, name: basename(path), type: "file" };
  try {
    const info = await lstat(path);
    if (info.isSymbolicLink()) {
      const target = await readlink(path);
      item.link_target = target;
      try {
        await stat(path);
      } catch {
        item.type = "broken_symlink";
        return item;
      }
      item.resolved = await realpath(path);
      item.type = (await isSkillDir(path)) ? "symlink_skill" : "symlink_non_skill";
    } else if (info.isDirectory()) {
      item.type = (await isSkillDir(path)) ? "skill_dir" : "missing_skill_md";
    } else {
      item.type = "file";
    }
    if (item.type === "skill_dir" || item.type === "symlink_skill") {
      item.content_hash = await contentHash(path);
    }
  } catch (error) {
    item.type = "error";
    item.error = error instanceof Error ? error.message : String(error);
  }
  return item;
}

async function exists(path: string): Promise<boolean> {
  try {
    await stat(path);
    return true;
  } catch {
    return false;
  }
}

export async function scanDir(path: string): Promise<ScanDirResult> {
  const result: ScanDirResult = { path, exists: await exists(path), entries: [] };
  if (!result.exists) {
    return result;
  }
  const info = await stat(path);
  if (!info.isDirectory()) {
    result.error = "not a directory";
    return result;
  }
  const entries = (await readdir(path)).sort();
  result.entries = await Promise.all(entries.map((entry) => classifyEntry(join(path, entry))));
  return result;
}

export type ScanResult = {
  locations: Record<string, ScanDirResult>;
  duplicates: {
    names: Record<string, string[]>;
    content_hashes: Record<string, string[]>;
  };
};

export async function scan(options: { project?: string; env?: Env } = {}): Promise<ScanResult> {
  const env = options.env ?? process.env;
  const locations: Record<string, string> = {
    inbox: inboxDir(env),
    store: skillsRoot(env),
    claude_global: adapter("claude", env).globalDir(),
    codex_global: adapter("codex", env).globalDir()
  };
  if (options.project) {
    locations.claude_project = adapter("claude", env).projectDir(options.project);
    locations.codex_project = adapter("codex", env).projectDir(options.project);
  }

  const scannedEntries = await Promise.all(Object.entries(locations).map(async ([name, path]) => [name, await scanDir(path)] as const));
  const scanned = Object.fromEntries(scannedEntries);
  const names: Record<string, string[]> = {};
  const hashes: Record<string, string[]> = {};

  for (const [locationName, location] of Object.entries(scanned)) {
    for (const entry of location.entries) {
      if (entry.type === "skill_dir" || entry.type === "symlink_skill") {
        (names[entry.name] ??= []).push(`${locationName}:${entry.path}`);
        if (entry.content_hash) {
          (hashes[entry.content_hash] ??= []).push(`${locationName}:${entry.path}`);
        }
      }
    }
  }

  return {
    locations: scanned,
    duplicates: {
      names: Object.fromEntries(Object.entries(names).filter(([, value]) => value.length > 1)),
      content_hashes: Object.fromEntries(Object.entries(hashes).filter(([, value]) => value.length > 1))
    }
  };
}
